text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,524 @@
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `core` libs changes
+
+The `Receiver` trait is made public (removing its `#[doc(hidden)])` attribute), exposing it under `core::ops`. It gains a `Target` associated type.
+
+This trait marks types that can be used as receivers other than the `Self` type of an impl or trait definition.
+
+```rust
+pub trait Receiver {
+    type Target: ?Sized;
+}
+```
+
+A blanket implementation is provided for any type that implements `Deref`:
+
+```rust
+impl<P: ?Sized, T: ?Sized> Receiver for P
+where
+    P: Deref<Target = T>,
+{
+    type Target = T;
+}
+```
+
+(See [alternatives](#no-blanket-implementation) for discussion of the tradeoffs here.)
+
+It is also implemented for `&T`, `&mut T`, `*const T` and `*mut T`.
+
+## Compiler changes
+
+The existing Rust [reference section for method calls describes the algorithm for assembling method call candidates](https://doc.rust-lang.org/reference/expressions/method-call-expr.html). This algorithm changes in one simple way: instead of dereferencing types (using the `Deref<Target=T>`) trait, we use the new `Receiver<Target=T>` trait to determine the next step.
+
+(Note that the existing algorithm isn't quite as simple as following the chain of `Deref<Target=T>`. In particular, `&T` and `&mut T` are considered as candidates too at each step; this RFC does not change that.)
+
+Because a blanket implementation is provided for users of the `Deref` trait and for `&T`/`&mut T`, the net behavior is similar. But this provides the opportunity for types which can't implement `Deref` to act as method receivers.
+
+Dereferencing a raw pointer usually needs `unsafe` (for good reason!) but in this case, no actual dereferencing occurs. This is used only to determine a list of method candidates; no memory access is performed and thus no `unsafe` is needed.
+
+## Object safety
+
+Receivers are object safe if they implement the (unstable) `core::ops::DispatchFromDyn` trait.
+
+As not all receivers might want to permit object safety or are unable to support it, object safety should remain being encoded in a different trait than the here proposed `Receiver` trait, likely `DispatchFromDyn`.
+
+This RFC does not propose any changes to `DispatchFromDyn`. Since `DispatchFromDyn` is unstable at the moment, object-safe receivers might be delayed until `DispatchFromDyn` is stabilized. `Receiver` is not blocked on further `DispatchFromDyn` work, since non-object-safe receivers already cover a big chunk of the use-cases.
+
+## Lifetime elision
+
+Arbitrary `self` parameters may involve lifetimes.
+
+This RFC currently proposes no changes to the standard lifetime elision rules. There are no known cases where ambiguity results. However, the opposite problem does apply: sometimes explicit lifetimes are required for methods with an arbitrary self type, even if they're not required for an equivalent free function. 
```

Comment:
This is still an ill-formed statement. The existing rules can't even apply to your new examples. You are extending the domain of legal function declarations that elision has to work on, you have to say how it behaves on that extended domain!

Reply:
> You are extending the domain of legal function declarations that elision has to work on

My assumption (which may be wrong) is that `self: Something` behaves identically to `any_other_param: Something`, which is why I don't see it as extending the domain of legal functions. It's just a parameter rename.

But, I'm going to proceed as if that assumption is wrong (some of the results in my example below suggest it is wrong) so I'll get my head around the existing lifetime elision code to explain what's up. This may take some time.

Reply:
That assumption is wrong. `fn foo(x: &T, y: &U) -> &Z` is invalid, but `fn foo(&self, y: &U) -> &Z` is allowed. The `self` parameter is special, if present then an elided lifetime in the return type will default to the lifetime of `self` even if there are other lifetimes in the signature.

Please have a look at https://doc.rust-lang.org/beta/reference/lifetime-elision.html :)

Reply:
Yep, thanks for working through this with me.

The current method-specific rule is:
> If the receiver has type `&Self` or `&mut Self`, then the lifetime of that reference to Self is assigned to all elided output lifetime parameters.

I had forgotten about that rule, but I think it's probably wise if it doesn't apply in this case. i.e. I propose that this rule continues to strictly apply _only_ to `&Self` and `&mut Self`, then no lifetime elision applies to `self: AnythingElse<T>` nor `self: &AnythingElse<T>` nor `self: &mut AnythingElse<T>`? That is, in the case of an arbitrary-self-type receiver, no elision occurs.

This feels like a case (rather like blocking generic `Receiver`s for now) where we can simplify the problem space for now, and later relax these rules if use-cases present themselves?

WDYT?

Reply:
I think that documentation is outdated, because this compiles:
```rust
use std::rc::Rc;

struct S;

impl S {
    fn test(self: Rc<&Self>) -> &i32 { todo!() }
}
```

Reply:
Ah OK. I'll probably aim to fix the current documentation (and thus come to understand the current behavior) before returning to this comment thread on this RFC. I'll leave this comment thread open but resolve the other two threads related to lifetime elision, since they're all related.

Reply:
Still digging but leaving notes on the current rules (in _stable_ Rust) because I'm finding them surprising.

With `struct Concrete(u32);` and all the following methods in an `impl Concrete` block.

Does not compile:
```rust
fn n(self: &Box<Self>, _other: &u32) -> &u32 {
    &self.0
}
```
This gives an error saying that we're returning data with the lifetime of `_other`; that is, the lifetime elision rules have not spotted the lifetime attached to `self`, and assume therefore that the output lifetime must belong to the other parameter. Adding an explicit lifetime solves this. Arguably this is a bug? Though I can't see when it would have actual bad consequences.

Also does not compile, and gives a curious message:
```rust
fn n1(self: &Box<Self>) -> &u32 {
    &self.0
}
```
Says "this function's return type contains a borrowed value, but the signature does not say which `self` it is borrowed from". Adding an explicit lifetime solves this.

That diagnostic, at least, feels non-optimal.

Does compile:

```rust
fn m(self: Box<&Self>, _other: &u32) -> &u32 {
    &self.0
}
fn g(self: Box<&Self>, _other: &u32) -> &u32 {
    &self.0
}
fn g1<'a, 'b>(self: &'a Box<&'b Self>, _other: &u32) -> &'a u32 {
    &self.0
}
fn g2<'a, 'b>(self: &'a Box<&'b Self>, _other: &u32) -> &'b u32 {
  &self.0
}
```

`g` seems to be ambiguous, although in practice Rust seems able to prove that lifetimes a and b are equivalent (hence both `g1` and `g2` compiling). It makes a difference to callers. `g` has the same semantics as `g2`.

Reply:
`'a` and `'b` are not equivalent. Two things are happening:
- the type of `self: &'a Box<&'b Self>` adds an implicit `'b: 'a` bound, which means that the inner value will stay live as long as `'a` is. This makes returning `&'a u32` correct.
- `self.0` copies a `&'b Self` out of the box, after which we don't need the box anymore. This allows us to return `&'b u32`.

It's not the same however: `g2` is more general than `g1` since `'a` could be much shorter than `'b`.

It seems lifetime elision is doing something like "take the lifetime of the innermost reference", which does infer the most general lifetime. It feels like a good general rule? It likely won't generalize to more general `Receiver`s though.

Reply:
> This gives an error saying that we're returning data with the lifetime of _other; that is, the lifetime elision rules have not spotted the lifetime attached to self, and assume therefore that the output lifetime must belong to the other parameter. Adding an explicit lifetime solves this. Arguably this is a bug? Though I can't see when it would have actual bad consequences.

That's a bug indeed, good catch!

Reply:
I have rewritten this section to say pretty much:

* There are already confusing/wrong lifetime situations in stable Rust today
* Please discuss them on the [bug report](https://github.com/rust-lang/rust/issues/117715) instead of here (though I give a preview of what we're currently intending to make the rules on that bug)
* We don't intend to change anything here in the RFC

So I'm going to resolve this thread for now. Thanks for all the discussion and guiding me towards the current faulty behavior.